text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,439 @@
+
+This proposal also targets documenting individual crates and their dependencies in non-cargo build systems. As will be explained, doc targets in non-cargo build systems often do not support cross-crate information.
+
+There are some files in the rustdoc output directory that are read and overwritten during every invocation of rustdoc. This proposal refers to these files as **cross-crate information**, or **CCI**, as in <https://rustc-dev-guide.rust-lang.org/rustdoc.html#multiple-runs-same-output-directory>.
+
+Build systems may run build actions in a distributed environment across separate logical filesystems. It might also be desirable to run rustdoc in a lock-free parallel mode, where every rustdoc process writes to a disjoint set of files.
+
+Cargo fully supports cross-crate information, at the cost of requiring global read-write access to the doc root (`target/doc`). There are significant scalability issues with this approach.
+
+Rustdoc needing global mutable access to the files that encode this cross-crate information has implications for caching, reproducible builds, and content hashing. By adding an option to avoid this mutation, rustdoc will serve as a first-class citizen in non-cargo build systems.
+
+These considerations motivate adding an option for outputting partial CCI (parts), which are merged (linked) with a later step.
+
+# Guide-level explanation
+
+In this example, there is a crate `t` which defines a trait `T`, and a crate `s` which defines a struct `S` that implements `T`. Our goal in this demo is for `S` to appear as an implementer in `T`'s docs, even if `s` and `t` are documented independently. This guide will be assuming that we want a crate `i` that serves as our documentation index. See the Unresolved questions section for ideas that do not require an index crate.
+
+```shell
+mkdir -p t/src s/src i/src merged/doc
+echo "pub trait T {}" > t/src/lib.rs
+echo "pub struct S; impl t::T for S {}" > s/src/lib.rs
+MERGED=file://$(realpath merged/doc)
+```
+
+[Actively use](https://doc.rust-lang.org/rustc/command-line-arguments.html#--extern-specify-where-an-external-library-is-located) `t` and `s` in `i`. The `extern crate` declarations are not needed if the crates are otherwise referenced in the index; intra-doc links are enough.
+
+```shell
+echo "extern crate t; extern crate s;" > i/src/lib.rs
+```
+
+Compile the crates.
+
+```shell
+rustc --crate-name=t --crate-type=lib --edition=2021 --emit=metadata --out-dir=t/target t/src/lib.rs
+rustc --crate-name=s --crate-type=lib --edition=2021 --emit=metadata --out-dir=s/target --extern t=t/target/libt.rmeta s/src/lib.rs
+```
+
+Document `s` and `t` independently, providing `--merge=none`, `--write-info-json`.
+
+```shell
+rustdoc \
+    -Z unstable-options \
+    --crate-name=t \
+    --crate-type=lib \
+    --edition=2021 \
+    --out-dir=t/target/doc \
+    --extern-html-root-url t=$MERGED \
+    --merge=none \
+    --write-info-json=t/target/doc.parts/t/crate-info.json \
```

Comment:
I wonder if we can come up with a more descriptive name than `doc.parts`? Maybe `doc.cache`, `doc.shared`, `doc.global`, or `doc.metadata`? IIUC, the name would only be hardcoded in the cargo implementation, but it'd still be good to have a clear name; also we should explicitly decide whether the name and structure of the directory would have any stability.

Reply:
Yeah, the implementation itself would be agnostic to the name of the directory. I think the structure of the directory should not be stable, in the same way that we agreed that the contents of crate-info.json should not bestable. `doc.parts` is meant to invoke the fact that it includes pieces / parts of files that are merged together. It also invokes the .part suffix for downloaded files, which communicates that `doc.parts` will not be readable until it's fully merged.